contracts/Vesting.sol

```diff
@@ -0,0 +1,137 @@
+            _amountOfDistribution
+        );
+        periodOfDistributionInDays = _periodOfDistributionInDays;
+        amountOfDistributionPerUser = _amountOfDistributionPerUser;
+        maxUsers = _amountOfDistribution / amountOfDistributionPerUser;
+    }
+
+    /// @notice this function can only be called by the owner of the contract
+    /// @dev call this function to start vesting
+
+    function join() public nonReentrant {
+        require(periodOfDistributionInDays != 0, "Vesting has not started");
+        infoOfUser[msg.sender].userTime = getCurrentTime();
+        require(
+            infoOfUser[msg.sender].claimedRevard == 0,
+            "You have already started vesting, use claimRevard to receive your revard"
+        );
+        users++;
+        require(
+            maxUsers >= users,
+            "Unfortunately, the tokens for distribution have already been distributed, or are distributing to the remaining users"
+        );
+        infoOfUser[msg.sender].finalDistributionTime =
+            getCurrentTime() +
+            (periodOfDistributionInDays * 86400);
+    }
+
+    /// @dev Returns current timestamp
+    function getCurrentTime() internal view virtual returns (uint256) {
+        return block.timestamp;
+    }
+
+    /// @notice call this function to receive available tokens
+    function claimRevard() public nonReentrant {
+        require(periodOfDistributionInDays != 0, "Vesting has not started");
+        require(
+            infoOfUser[msg.sender].userTime > 0,
+            "You have not joined the vesting"
+        );
+        require(
+            infoOfUser[msg.sender].finalDistributionTime != 0,
+            "You have received all tokens"
+        );
+
+        uint256 tokenPerSecond = (amountOfDistributionPerUser * 1000000) /
+            (periodOfDistributionInDays * 86400);
+        infoOfUser[msg.sender].claimedRevard =
+            (getCurrentTime() - infoOfUser[msg.sender].userTime) *
+            tokenPerSecond;
+
+        if (getCurrentTime() >= infoOfUser[msg.sender].finalDistributionTime) {
+            infoOfUser[msg.sender].claimedRevard =
+                amountOfDistributionPerUser -
+                infoOfUser[msg.sender].allClaimedRevard;
+            _token.safeTransfer(
+                msg.sender,
+                infoOfUser[msg.sender].claimedRevard
+            );
+            infoOfUser[msg.sender].finalDistributionTime = 0;
+        } else {
+            _token.safeTransfer(
+                msg.sender,
+                infoOfUser[msg.sender].claimedRevard / 1000000
+            );
+            infoOfUser[msg.sender].userTime = getCurrentTime();
+        }
+        infoOfUser[msg.sender].allClaimedRevard =
+            infoOfUser[msg.sender].allClaimedRevard +
+            infoOfUser[msg.sender].claimedRevard /
+            1000000;
+    }
+
+    /// @notice Use this function to see, what revard you can claim
+    function showRevardToClaim() public view returns (uint256) {
```

Comment:
If u have function to get current reward, can't u just use this function to calculate reward and let user claim it?